### PR TITLE
Add `clientTraceMetadata` experimental option to propagate tracing data to the client

### PR DESCRIPTION
**Pull request by lforst**
### What?

This PR adds an experimental option `clientTraceMetadata` that will use the existing OpenTelemetry functionality to propagate conventional OpenTelemetry trace information to the client.

The propagation metadata is propagated to the client via meta tags, having a `name` and a `content` attribute containing the value of the tracing value:

```html
<html>
    <head>
        <meta name="baggage" content="key1=val1,key2=val2">
        <meta name="traceparent" content="00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01">
        <meta name="custom" content="foobar">
    </head>
</html>
```

The implementation adheres to OpenTelemetry as much as possible, treating the meta tags as if they were tracing headers on outgoing requests. The `clientTraceMetadata` will contain the keys of the metadata that're going to injected for tracing purpose.

### Why?

Telemetry providers usually want to provide visibility across the entire stack, meaning it is useful for users to be able to associate, for example, web vitals on the client, with a span tree on the server. In order to be able to correlate tracing events from the front- and backend, it is necessary to share something like a trace ID or similar, that the telemetry providers can pick up and stitch back together to create a trace.

### How?

The tracer was extended with a method `getTracePropagationData()` that returns the propagation data on the currently active OpenTelemetry context. 
We are using `makeGetServerInsertedHTML()` to inject the meta tags into the HTML head for dynamic requests.
The meta tags are generated through using the newly added `getTracePropagationData()` method on the tracer.

It is important to mention that **the trace information should only be propagated for the initial loading of the page, including hard navigations**. Any subsequent operations should not propagate trace data from the server to the client, as the client generally is the root of the trace. The exception is initial pageloads, since while the request starts on the client, no JS has had the opportunity to run yet, meaning there is no trace propagation on the client before the server hasn't responded.

Situations that we do not want tracing information to be propagated from the server to the client:
- _Prefetch requests._ Prefetches generally start on the client and are already instrumented.
- _Any sort of static precomputation, including PPR._ If we include trace information in static pages, it means that all clients that will land on the static page will be part of the "precomputation" trace. This would lead to gigantic traces with a ton of unrelated data that is not useful. The special case is dev mode where it is likely fine to propagate trace information, even for static content, since it is usually not actually static in dev mode.
- _Clientside (soft) navigations._ Navigations start on the client and are usually already instrumented.

### Alternatives considered

An implementation that purely lives in user-land could have been implemented with `useServerInsertedHTML()`, however, that implementation would be cumbersome for users to set up, since the implementation of tracing would have to happen in a) the instrumentation hook, b) in a client-component that is used in a top-level layout.

### Related issues/discussions

- https://github.com/vercel/next.js/issues/47660
- https://github.com/vercel/next.js/discussions/62353 (Could be used as an alternative to the server-timing header)
- https://github.com/getsentry/sentry-javascript/issues/9571

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: cd2c1ec68fb98ebb06a94d47ebfd126642c9bb38

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by lforst** [comment]
I am happy to add more tests! I'd like to gauge the appetite for something like this.

File: test/e2e/client-trace-metadata/app/layout.tsx [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export default function Root({ children }: { children: React.ReactNode }) {
```
Comment (dvoytenko) [inline test/e2e/client-trace-metadata/app/layout.tsx:1]: It'd be good to combine these tests with the `e2e/opentelemetry` tests.

**Review by dvoytenko — DISMISSED** [review]

File: packages/next/src/server/lib/trace/tracer.ts [hunk 2]
```diff
@@ -172,6 +193,13 @@ class NextTracerImpl implements NextTracer {
     return context
   }
 
+  public getTracePropagationData(): ClientTraceDataEntry[] {
+    const activeContext = context.active()
```
[the remark this answers is not in the document]
Comment (lforst) [reply packages/next/src/server/lib/trace/tracer.ts:197]: In general, it is beneficial to propagate sampled and un-sampled contexts because telemetry systems often do traced-based sampling, where a tracing decision for a particular trace (usually decided on the first service) is propagated across the entire system, and the individual services of the system only report spans, if the incoming tracing decision is positive.

For example in the `traceparent` header there is a flag indicating whether the trace is sampled or not: https://www.w3.org/TR/trace-context/#sampled-flag If we didn't propagate the data when the trace is not sampled, the flag would not be needed in the first place.

Does that make sense? All in all this is good because it allows us to have less logic here. :)

[the remark this answers is not in the document]
Comment (dvoytenko) [reply packages/next/src/server/lib/trace/tracer.ts:197]: That's what I meant - the absence of the trace can be used to mean the traceflags are `00`. But I agree with the overall reasoning. You can keep this.

**Comment by lforst** [comment]
@dvoytenko Re our slack convo: To clarify in what cases trace metadata should be propagated to the client:

- **tldr:** Effectively only for pageloads or hard navigations.
- **Not** for any statically generated pages, as the "cached" trace data would be picked up by all clients and create huge and unusable traces.
- **Not** for any PPR content. (same reason as static pages)
- **Not** for soft navigations. Traces for navigations originate from the client and do not need to be propagated back.

I added tests for all of the above.

I think moving forward we need to get clearance two things:
- The naming of the meta tags. Imo, this can be `_next-trace-data-*` for now as the option is experimental and we can update later if we find a better name.
- Whether what I did in `makeGetServerInsertedHTML()` makes sense or whether this needs to live somewhere else.

**Comment by dvoytenko** [comment]
> * **Not** for any statically generated pages, as the "cached" trace data would be picked up by all clients and create huge and unusable traces.
> * **Not** for any PPR content. (same reason as static pages)

This sounds reasonable. But how does your code avoid having these values produced in these modes?

P.S. We actually produce trace references for both static and PPR. But we are not using them as parent traces - that would indeed make the parent traces huge and not useful. We simply use them as links/tags for other traces to add. It reads like this: "this trace was produced by a page which was originally statically generated".

**Comment by lforst** [comment]
@dvoytenko Thanks for giving this another pass!

> > * **Not** for any statically generated pages, as the "cached" trace data would be picked up by all clients and create huge and unusable traces.
> > * **Not** for any PPR content. (same reason as static pages)
> 
> This sounds reasonable. But how does your code avoid having these values produced in these modes?

Currently, we don't explicitly check for PPR/SSG, but I _believe_ the code path where we would generate the `<meta>` tags is simply not executed for PPR and static pages. In theory, the tests I added verify this 🤔 ([1](https://github.com/vercel/next.js/pull/64256/files#diff-c8dda8307d78c09f1f8b30c67ca78de6d5982132e6c161ea041e41ad7556cf97R48), [2](https://github.com/vercel/next.js/pull/64256/files#diff-c8dda8307d78c09f1f8b30c67ca78de6d5982132e6c161ea041e41ad7556cf97R82), [3](https://github.com/vercel/next.js/pull/64256/files#diff-c8dda8307d78c09f1f8b30c67ca78de6d5982132e6c161ea041e41ad7556cf97R11)). Maybe someone working on the Next.js team working on this part of the code can confirm. If you feel like it's a good idea, I can also explicitly check for SSG and PPR and not inject the meta tags, however, I would lean towards not doing that because it would essentially be dead code assuming the code path is actually not hit.

The point about generating traces for static content also makes complete sense 👍 I assume you're using that already to track Next.js compile speed? We should just make sure that those traces do not leak into production monitoring, but I think we are aligned on that point and the tests already test for it.

File: packages/next/src/server/app-render/app-render.tsx [hunk 3]
```diff
@@ -947,6 +947,9 @@ async function renderToHTMLOrFlightImpl(
         renderServerInsertedHTML,
         serverCapturedErrors: allCapturedErrors,
         basePath: renderOpts.basePath,
+        traceData: renderOpts.experimental.clientTraceMetadata
```
[the remark this answers is not in the document]
Comment (lforst) [reply packages/next/src/server/app-render/app-render.tsx:950]: I put this feature behind an experimental flag for two reasons:
- This feature _is_ a bit experimental and maybe until we iron out some issues, it makes sense to put it behind an experiment.
- I wanted to make this feature opt-in because if people already have an OpenTelemetry propagator registered, the tracing data will go to the client and potentially leak data. It is unlikely, but it may happen.

If you feel like this is an acceptable risk, I am more than happy to remove the feature flag because it would drastically improve the out-of-the-box tracing experience with Next.js!

File: packages/next/src/server/app-render/make-get-server-inserted-html.tsx [hunk 4]
```diff
@@ -81,6 +84,14 @@ export function makeGetServerInsertedHTML({
               return <script key={polyfill.src} {...polyfill} />
             })
         }
+        {traceData.map(({ key, value }) => {
```
[the remark this answers is not in the document]
Comment (lforst) [reply packages/next/src/server/app-render/make-get-server-inserted-html.tsx:87]: I am not fully sure if I understand your concern correctly, but I think this is already supported.

If you have multiple OpenTelemetry propagation keys that have the same value, they will already be added as multiple `<meta />` tags. [Here we generate the array](https://github.com/vercel/next.js/pull/64256/files#diff-27f951b8613914f9b7e5a8a8a8bbf4a151bf1202f7e290a46e5ed5480567f138R187-R192), and right where you commented, we render the tags.

If your comment is about having additional attributes on the rendered `<meta />` tags, I don't think that is needed because OpenTelemetry trace data is just key-value pairs.

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| buildDuration | 19.7s | 17.5s | N/A |
| buildDurationCached | 9.3s | 8.2s | N/A |
| nodeModulesSize | 238 MB | 238 MB | ⚠️ +40.5 kB |
| nextStartRea..uration (ms) | 433ms | 420ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.5 kB | 31.5 kB | N/A |
| 3304.HASH.js gzip | 169 B | 169 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.09 kB | 5.09 kB | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 228 B | 227 B | N/A |
| main-HASH.js gzip | 31.5 kB | 31.6 kB | N/A |
| webpack-HASH.js gzip | 1.64 kB | 1.65 kB | N/A |
| Overall change | 50.5 kB | 50.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 194 B | N/A |
| _error-HASH.js gzip | 193 B | 191 B | N/A |
| amp-HASH.js gzip | 510 B | 510 B | ✓ |
| css-HASH.js gzip | 342 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.51 kB | 2.52 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 365 B | 364 B | N/A |
| hooks-HASH.js gzip | 389 B | 391 B | N/A |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 269 B | 268 B | N/A |
| link-HASH.js gzip | 2.68 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 326 B | N/A |
| script-HASH.js gzip | 395 B | 397 B | N/A |
| withRouter-HASH.js gzip | 323 B | 323 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.2 kB | 1.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| _buildManifest.js gzip | 482 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| edge-ssr.js gzip | 108 kB | 108 kB | N/A |
| page.js gzip | 3.05 kB | 3.05 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 625 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 27.8 kB | 27.8 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | N/A |
| app-page-exp..prod.js gzip | 97.6 kB | 97.7 kB | N/A |
| app-page-tur..prod.js gzip | 99.4 kB | 99.5 kB | N/A |
| app-page-tur..prod.js gzip | 93.6 kB | 93.7 kB | N/A |
| app-page.run...dev.js gzip | 145 kB | 145 kB | N/A |
| app-page.run..prod.js gzip | 92.1 kB | 92.2 kB | N/A |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.2 kB | 21.2 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| server.runti..prod.js gzip | 65.3 kB | 65.3 kB | N/A |
| Overall change | 197 kB | 197 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lforst/nextjs-fork lforst-propagate-trace-to-client | Change |
| - | - | - | - |
| 0.pack gzip | 1.61 MB | 1.61 MB | N/A |
| index.pack gzip | 106 kB | 107 kB | ⚠️ +694 B |
| Overall change | 106 kB | 107 kB | ⚠️ +694 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1552: /***/ (
+    /***/ 4070: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(5237);
+          return __webpack_require__(396);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 2016: /***/ (module, exports, __webpack_require__) => {
+    /***/ 8490: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(422)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6074)
+        __webpack_require__(2457)
       );
-      const _getimgprops = __webpack_require__(9571);
-      const _imageconfig = __webpack_require__(6567);
-      const _imageconfigcontextsharedruntime = __webpack_require__(419);
-      const _warnonce = __webpack_require__(4486);
-      const _routercontextsharedruntime = __webpack_require__(162);
+      const _getimgprops = __webpack_require__(7932);
+      const _imageconfig = __webpack_require__(5706);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9483);
+      const _warnonce = __webpack_require__(9035);
+      const _routercontextsharedruntime = __webpack_require__(4829);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -379,7 +379,7 @@
       /***/
     },
 
-    /***/ 9571: /***/ (
+    /***/ 7932: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -395,9 +395,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(4486);
-      const _imageblursvg = __webpack_require__(133);
-      const _imageconfig = __webpack_require__(6567);
+      const _warnonce = __webpack_require__(9035);
+      const _imageblursvg = __webpack_require__(2642);
+      const _imageconfig = __webpack_require__(5706);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -772,7 +772,7 @@
       /***/
     },
 
-    /***/ 133: /***/ (__unused_webpack_module, exports) => {
+    /***/ 2642: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -827,7 +827,7 @@
       /***/
     },
 
-    /***/ 4085: /***/ (
+    /***/ 503: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -854,10 +854,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(7456);
-      const _getimgprops = __webpack_require__(9571);
-      const _imagecomponent = __webpack_require__(2016);
+      const _getimgprops = __webpack_require__(7932);
+      const _imagecomponent = __webpack_require__(8490);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -889,7 +889,7 @@
       /***/
     },
 
-    /***/ 6996: /***/ (__unused_webpack_module, exports) => {
+    /***/ 7240: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -924,7 +924,7 @@
       /***/
     },
 
-    /***/ 5237: /***/ (
+    /***/ 396: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -941,8 +941,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(1527);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
-      var next_image = __webpack_require__(1577);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
+      var next_image = __webpack_require__(73);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -972,12 +972,8 @@
       /***/
     },
 
-    /***/ 1577: /***/ (
-      module,
-      __unused_webpack_exports,
-      __webpack_require__
-    ) => {
-      module.exports = __webpack_require__(4085);
+    /***/ 73: /***/ (module, __unused_webpack_exports, __webpack_require__) => {
+      module.exports = __webpack_require__(503);
 
       /***/
     },
@@ -987,7 +983,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1552)
+      __webpack_exec__(4070)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>2453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 9188d1bb6f1bf9b094460fb0ce71fe92e4adfd27

File: packages/next/src/server/app-render/make-get-server-inserted-html.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (huozhi) [reply packages/next/src/server/app-render/make-get-server-inserted-html.tsx:87]: I was checking what will the `getTracePropagationData` can be used, seems not all of them will be used for rendering injected metadata. Some could be used in request/response headers. Maybe we should give everything a prefix for categorizing. Such as `<category>@<key>:<valye>`

To inject metadata rendering tags we can use  `metadata@meta-tag-key:the-value` for identify it's for metadata and we only filter the ones starting with `metadata@` for injecting meta tags.

If we need to use propagation data for other cases, we can add a different category, such as `headers@key:value`

[the remark this answers is not in the document]
Comment (lforst) [reply packages/next/src/server/app-render/make-get-server-inserted-html.tsx:87]: Trying to sum up my thoughts (and also our slack convo):

Long story short, I think adding something like a `namespace@` would not work for the use case I was attempting to fix with this PR and would be mixing too many concepts.

Generally speaking, OpenTelemetry users have a convention for the name of the keys in their propagation data. Having a `metadata@` prefix would mess with that convention. When implementing an OpenTelemetry `Propagator`, you cannot know at runtime whether your data will be `inject`ed into a meta tag or attached as headers to outgoing requests. Having the `metadata@` prefix in headers of outgoing requests would certainly be problematic in many cases.

I do agree with the fact that Next.js probably needs a way to dynamically attach response headers. However, I don't think using OpenTelemetry is the right vehicle for that. Returning trace information to the client as response headers is generally of no use, because you cannot extract it with JS (except for very few exceptions like `Server-Timing`). I suggest that if you guys want to add a mechanism to return response headers you add it through a more explicit API.

**Review by huozhi — APPROVED** [review]

**Comment by huozhi** [comment]
After discussing with @dvoytenko, we decided to change `clientTraceMetadata` to a list that represents the names of metadata that are going to be propagated to in the client. I also removed the `_next-trace-data-` prefix so the 3p can still rely on the standard fields such as `traceparent`, `tracestate` if they're parsing them from html page

**Comment by MrLoh** [comment]
Is it known yet, what next.js release this will be included in?
